### ISSUE TITLE
debugging blazor application in non-development mode does not work

### ISSUE THREAD

**Issue by pavlexander** [dotnet/aspnetcore#30685]
### Describe the bug
When I debug my application with `Development` environment variable set - everything works. When I remove the variable or rename it to `Production`, then website becomes unreachable.

### To Reproduce
1. create a new `Blazor WASM` application with `asp.net core hosting` option selected
2. change the startup mode from `IIS Express` to Kestrel

![image](https://user-images.githubusercontent.com/20644772/110083224-1ae9dd80-7d97-11eb-8382-1bfb09b2366e.png)

3. start debug session.

**Expected, actual:** browser window launched, website is reachable/loaded.

4. go to launch settings and remove the `Development` value or set it to `Production` for `ASPNETCORE_ENVIRONMENT` env variable setting

![image](https://user-images.githubusercontent.com/20644772/110083190-0efe1b80-7d97-11eb-8d21-0408060c42e5.png)

5. start debug session.

#### Expected

browser window launched, website is reachable/loaded.

#### Actual

![image](https://user-images.githubusercontent.com/20644772/110082616-43250c80-7d96-11eb-9424-bd4de5979cd3.png)

* browser window opened but the default webpage is not opened

* if you manually navigate to either of the pages, i.e. `http://localhost:5000`, you will see the **This localhost page can’t be found** (404) error

![image](https://user-images.githubusercontent.com/20644772/110083080-e5dd8b00-7d96-11eb-8b88-f978409e9a32.png)

* if you wait _exactly 1 minute_ with debug session running and staring at window you will see the exception in `VisualStudio` - **failed to launch debug adapter**:

![image](https://user-images.githubusercontent.com/20644772/110082845-94cd9700-7d96-11eb-96b7-fa8b6985f0a1.png)

### Exceptions (if any)

1. This localhost page can’t be found
2. failed to launch debug adapter

### Further technical details

```
Microsoft Visual Studio Community 2019
Version 16.8.6
VisualStudio.16.Release/16.8.6+31019.35
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Community

Visual C++ 2019   00435-60000-00000-AA596
Microsoft Visual C++ 2019

ASP.NET and Web Tools 2019   16.8.560.26713
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2052803+84e121f1403378489b842e1797df2f3f5a49ac3c
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.8.560.26713
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.8.560.26713
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.8.560.26713
Azure Functions and Web Jobs Tools

C# Tools   3.8.0-5.20604.10+9ed4b774d20940880de8df1ca8b07508aa01c8cd
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Edit Project   1.6.34
An open source Visual Studio extension to add the context menu for editing project/solution file.

GitExtensions   1.0
Git Extensions is a graphical user interface for Git that allows you to control Git without using the command-line

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Markdown Editor   1.12.253
A full featured Markdown editor with live preview and syntax highlighting. Supports GitHub flavored Markdown.

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.30924.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.113+g422d40002e.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   5.8.1
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/
```

```
.NET SDK (reflecting any global.json):
 Version:   5.0.103
 Commit:    72dec52dbd

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.103\

Host (useful for support):
  Version: 5.0.3
  Commit:  c636bbdc8a

.NET SDKs installed:
  5.0.103 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.WindowsDesktop.App 5.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

## additional options tried

I tried changing

`        "applicationUrl": "https://localhost:5001;http://localhost:5000",`
to

`        "applicationUrl": "http://localhost:5000",`

commenting out HTTPS redirect: `//app.UseHttpsRedirection();` in `Startup.cs`
removing non-development specific configuration:

```
            //if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.UseWebAssemblyDebugging();
            }
            /*else
            {
                app.UseExceptionHandler("/Error");
                // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                app.UseHsts();
            }

            //app.UseHttpsRedirection();*/
```

Neither of my attempts helped.. I am still not able to debug the application no matter what I do... Am I doing something wrong?

**Comment by pavlexander** [comment]
**upd** Seems like after some rebuilds and changes here and there I somehow managed to get the Chrome **debug adapter** related error go away.. But essentially the same error still persists - unable to open/debug the website:

![image](https://user-images.githubusercontent.com/20644772/110085532-e297ce80-7d99-11eb-93e1-85658509f055.png)

as you see, for simplification reasons I have completely disposed of HTTPS related code, so only HTTP binding supposed to work. But it isn't. Please refer to reproduction steps in original post for "clean", |barebone code| to repro the issue.

**Comment by javiercn** [comment]
@pavlexander thanks for contacting us.

The debugger proxy is only enabled during development by default. I would suggest you check your startup file and move it elsewhere if you want to test it on a different environment. See [here](https://github.com/dotnet/aspnetcore/blob/main/src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs#L99)

**Comment by pavlexander** [comment]
> The debugger proxy is only enabled during development by default. I would suggest you check your startup file and move it elsewhere if you want to test it on a different environment. See [here](https://github.com/dotnet/aspnetcore/blob/main/src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs#L99)

Are you perhaps talking about the `app.UseWebAssemblyDebugging();` extension method that adds debugging capabilities?

As I have mentioned in the original description - this was one of the things I have tried doing - increasing the scope of this extension method, like this:

![image](https://user-images.githubusercontent.com/20644772/110112678-ba6c9780-7dba-11eb-93a2-510fd0bf92c9.png)

It does not really solve the problem..

**Comment by javiercn** [comment]
@pavlexander thanks for the additional details.

I missed that bit in my original analysis. We'll take a look and update the issue accordingly.

**Comment by xendren** [comment]
Is there any way to escalate this or at least confirm if it only relates to debugging?  I'm trying to evaluate whether our company should use Blazor, Mudblazor, or choose something else.  Unfortunately, I have just encountered this bug with the stock Mudblazor and Blazor project templates from "dotnet new".  All you have to do to make it fail is to change the "ASPNETCORE_ENVIRONMENT" variable to something other than "**Development**".  With MudBlazor, you get a stack trace and the index page only partially renders with the mudblazor icon enormous.  With the template Blazor app, there are no exceptions raised, but the stock side panel fails to render.  

I tried removing the isDevelopment() logic completely in Program.cs, but I get the same behavior.  Somewhere in the bowls of the code, something is expecting that environment variable to equal Development, or it misfires.  
I'm using the latest .Net 6 SDK.  

Can someone please at least confirm this is ONLY an issue with debugging with Blazor before we start developing?  I don't want to choose Blazor and end up wasting our time later if we have to bypass the builtin .Net settings functionality to get it to work.

**Comment by mkArtakMSFT** [comment]
@thaystg any thoughts regarding this one? I'm not sure if this has been fixed already or if you'd like to move this to your repo. If so, feel free to transfer to the runtime repo.

**Comment by thaystg** [comment]
> @thaystg any thoughts regarding this one? I'm not sure if this has been fixed already or if you'd like to move this to your repo. If so, feel free to transfer to the runtime repo.

@maraf is this related to the other one, right?

**Comment by thaystg** [comment]
I cannot reproduce it @mkArtakMSFT

**Comment by hatefk** [comment]
Any update On this? 
While we have a couple of work arounds in our back-pocket it would be nice to use this method. It has been blocking us from releasing to Production,

**Comment by javiercn** [comment]
We don’t think this issue applies to newer Blazor versions. 

There are certain features that are only wired-up during development. If you are running in an environment different than development, you must turn those features on explicitly.

In particular:
- Static web assets (for referenced/RCL assets and webassembly files): builder.WebHost.UseStaticWebAssets();
- WebAssembly debugging (hosted WASM only): app.UseWebAssemblyDebugging(); 

Everything else either already behaves the same outside Development or isn’t required for the app to run. Please give it a try a supported .NET version and let us know if you still run into issues by filing a separate issue.
